sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java

```diff
@@ -385,43 +399,68 @@ public void process(
-                          runAsyncIgnoreFailure(
-                              closeWriterExecutor,
-                              () -> {
-                                // Remove the pin that is "owned" by the cache.
-                                client.unpin();
-                                client.close();
-                              }));
-              if (createAppendClient) {
-                info = info.createAppendClient(datasetService, getOrCreateStream, false);
-                // This pin is "owned" by the cache.
-                Preconditions.checkStateNotNull(info.streamAppendClient).pin();
-              }
-              return info;
-            } catch (Exception e) {
-              throw new RuntimeException(e);
-            }
-          };
-
       AtomicReference<AppendClientInfo> appendClientInfo =
           new AtomicReference<>(
-              APPEND_CLIENTS.get(element.getKey(), () -> getAppendClientInfo.apply(true)));
+              APPEND_CLIENTS.get(
+                  element.getKey(),
+                  () -> {
+                    @Nullable TableSchema tableSchema;
+                    if (autoUpdateSchema && updatedSchema.read() != null) {
+                      // We've seen an updated schema, so we use that.
+                      tableSchema = updatedSchema.read();
+                    } else {
+                      // Start off with the base schema. As we get notified of schema updates, we
+                      // will update the
+                      // descriptor.
+                      tableSchema =
+                          messageConverters
+                              .get(element.getKey().getKey(), dynamicDestinations, datasetService)
+                              .getTableSchema();
+                    }
+                    AppendClientInfo info =
+                        AppendClientInfo.of(
+                                Preconditions.checkStateNotNull(tableSchema),
+                                // Make sure that the client is always closed in a different thread
+                                // to
+                                // avoid blocking.
+                                client ->
+                                    runAsyncIgnoreFailure(
+                                        closeWriterExecutor,
+                                        () -> {
+                                          // Remove the pin that is "owned" by the cache.
+                                          client.unpin();
+                                          client.close();
+                                        }))
+                            .withAppendClient(datasetService, getOrCreateStream, false);
+                    // This pin is "owned" by the cache.
+                    Preconditions.checkStateNotNull(info.getStreamAppendClient()).pin();
+                    return info;
+                  }));
+      TableSchema updatedSchemaValue = updatedSchema.read();
```

Comment:
updatedSchema can be null?

Reply:
updatedSchema is a state variable. If it has never been set, then it will return null.